### PR TITLE
Fix EMA for multi-gpu training in the unconditional example

### PR DESCRIPTION
**Pull request by anton-l**
Now the unconditional EMA wrapper mimics the `EMAModel` from `train_text_to_image.py`. This isn't a full copy because the unconditional example uses a different decay schedule. 

Fixes broken training on multiple GPUs: #1772 #1895

### CODE REVIEW

File: examples/unconditional_image_generation/train_unconditional.py [hunk 1]
```diff
@@ -156,7 +156,6 @@ def parse_args():
     parser.add_argument(
         "--use_ema",
         action="store_true",
-        default=True,
         help="Whether to use Exponential Moving Average for the final model weights.",
     )
```
Comment (anton-l) [inline examples/unconditional_image_generation/train_unconditional.py:160]: Also removing the wrong default for consistency with other scripts (issue #1654)

Comment (patrickvonplaten) [reply examples/unconditional_image_generation/train_unconditional.py:160]: small breaking change but think that's fine!

File: src/diffusers/training_utils.py [hunk 2]
```diff
@@ -66,7 +68,12 @@ def __init__(
             power (float): Exponential factor of EMA warmup. Default: 2/3.
             min_value (float): The minimum EMA decay rate. Default: 0.
         """
-
+        deprecation_message = (
```
Comment (patrickvonplaten) [inline src/diffusers/training_utils.py:71]: Ah I didn't know that we already had a class here.

There is a relatively high chance that the EMA model will be moved here eventually if other training scripts use it. Should we maybe just update it now with a breaking change as it would be cleaner? If not:

- We have a deprecated EMAModel in `src/diffusers` that some people still use
- Two EMAModel in examples
- A big chance that we need to overwrite the deprecated model with the example one soon

=> Maybe easier to just do a breaking change now? Wdyt @patil-suraj @anton-l 

Not super happy about the two classes situation

Comment (patrickvonplaten) [reply src/diffusers/training_utils.py:71]: Ping here @anton-l @patil-suraj

Comment (anton-l) [reply src/diffusers/training_utils.py:71]: @patil-suraj feel free to take over if that's the case! I've tried to make unconditional EMAModel behave similarly to the one you've added to text2img. However, I'm not sure how easy it is to replace the integrated EMAModel without breaking changes: it has a slightly different API to the script counterparts

Comment (patil-suraj) [reply src/diffusers/training_utils.py:71]: Awesome, Will try to see if we can support both EMA warmup schedules cleanly.

Comment (patil-suraj) [reply src/diffusers/training_utils.py:71]: Will push the changes in this PR.

File: src/diffusers/training_utils.py [hunk 3]
```diff
@@ -39,88 +40,173 @@ def set_seed(seed: int):
...
-        power=2 / 3,
-        min_value=0.0,
-        max_value=0.9999,
-        device=None,
+        parameters: Iterable[torch.nn.Parameter],
+        decay: float = 0.9999,
+        update_after_step: int = 0,
+        use_ema_warmup: bool = False,
+        inv_gamma: Union[float, int] = 1.0,
+        power: Union[float, int] = 2 / 3,
```
Comment (patil-suraj) [inline src/diffusers/training_utils.py:56]: supported both ema schedules here; the only change is I removed the `min_value` and `max_value` and just added `decay`.

File: src/diffusers/training_utils.py [hunk 4]
```diff
@@ -39,88 +40,173 @@ def set_seed(seed: int):
...
         """
         Compute the decay factor for the exponential moving average.
         """
         step = max(0, optimization_step - self.update_after_step - 1)
-        value = 1 - (1 + step / self.inv_gamma) ** -self.power
+
+        if self.use_ema_warmup:
+            cur_decay_value = 1 - (1 + step / self.inv_gamma) ** -self.power
+        else:
+            cur_decay_value = (1 + step) / (10 + step)
```
Comment (patil-suraj) [inline src/diffusers/training_utils.py:95]: this is the default which is used for SD.

File: src/diffusers/training_utils.py [hunk 5]
```diff
@@ -39,88 +40,173 @@ def set_seed(seed: int):
...
+
+        self.optimization_step += 1
+
+        # Compute the decay factor for the exponential moving average.
+        decay = self.get_decay(self.optimization_step)
+        one_minus_decay = 1 - decay
+
+        for s_param, param in zip(self.shadow_params, parameters):
+            if param.requires_grad:
+                s_param.sub_(one_minus_decay * (s_param - param))
```
Comment (patil-suraj) [inline src/diffusers/training_utils.py:114]: this is equivalent to doing 

```python
ema_param.mul_(self.decay)
ema_param.add_(param.data.to(dtype=ema_param.dtype), alpha=1 - self.decay)
```

File: src/diffusers/training_utils.py [hunk 6]
```diff
@@ -39,88 +40,173 @@ def set_seed(seed: int):
...
             gamma=1, power=3/4 for models you plan to train for less (reaches decay factor 0.999 at 10K steps, 0.9999
             at 215.4k steps).
-
-        Args:
-            inv_gamma (float): Inverse multiplicative factor of EMA warmup. Default: 1.
-            power (float): Exponential factor of EMA warmup. Default: 2/3.
-            min_value (float): The minimum EMA decay rate. Default: 0.
         """
+        parameters = list(parameters)
+        self.shadow_params = [p.clone().detach() for p in parameters]
```
Comment (anton-l) [inline src/diffusers/training_utils.py:75]: The class expected a model, but now it expects a list of parameters. Can we somehow avoid this breaking change?

Comment (patil-suraj) [reply src/diffusers/training_utils.py:75]: Aah, good catch! IMO it's safer to just pass `parameters`,which makes it compatible with multi-gpu, deepspeed etc.
Think as @patrickvonplaten said, the breaking change should be fine.

File: examples/unconditional_image_generation/train_unconditional.py [hunk 7]
```diff
@@ -342,19 +342,35 @@ def transforms(examples):
...
+            inv_gamma=args.ema_inv_gamma,
+            power=args.ema_power,
+        )
+
     model, optimizer, train_dataloader, lr_scheduler = accelerator.prepare(
         model, optimizer, train_dataloader, lr_scheduler
     )
-    accelerator.register_for_checkpointing(lr_scheduler)
+    if args.use_ema:
+        accelerator.register_for_checkpointing(ema_model, lr_scheduler)
```
Comment (pcuenca) [inline examples/unconditional_image_generation/train_unconditional.py:358]:
```suggestion
        accelerator.register_for_checkpointing(ema_model)
```

Not required as verified by @patil-suraj in https://github.com/huggingface/diffusers/pull/1834#discussion_r1059397718

File: examples/unconditional_image_generation/train_unconditional.py [hunk 8]
```diff
@@ -342,19 +342,35 @@ def transforms(examples):
...
+    elif accelerator.mixed_precision == "bf16":
+        weight_dtype = torch.bfloat16
 
-    ema_model = EMAModel(
-        accelerator.unwrap_model(model),
-        inv_gamma=args.ema_inv_gamma,
-        power=args.ema_power,
-        max_value=args.ema_max_decay,
-    )
+    # Move text_encode and vae to gpu and cast to weight_dtype
```
Comment (pcuenca) [inline examples/unconditional_image_generation/train_unconditional.py:368]:
```suggestion
    # Move model to gpu and cast to weight_dtype
```

File: src/diffusers/training_utils.py (hunk 4 above)
Comment (pcuenca) [reply src/diffusers/training_utils.py:95]: Are the two previous implementations equivalent to this new one or are we overriding one of them? (Sorry, it's not clear to me from the diff).

File: src/diffusers/training_utils.py [hunk 9]
```diff
@@ -39,88 +40,173 @@ def set_seed(seed: int):
...
+        if not isinstance(self.use_ema_warmup, bool):
+            raise ValueError("Invalid use_ema_warmup")
+
+        self.inv_gamma = state_dict["inv_gamma"]
+        if not isinstance(self.inv_gamma, (float, int)):
+            raise ValueError("Invalid inv_gamma")
+
+        self.power = state_dict["power"]
+        if not isinstance(self.power, (float, int)):
+            raise ValueError("Invalid power")
```
[the remark this answers is not in the document]
Comment (pcuenca) [reply src/diffusers/training_utils.py:197]: I'd say we default to the values previously used in `train_text_to_image.EMAModel`, since we know what they are. Also, the previous EMA model in `training_utils` could not be serialized anyway (as far as I can tell), so we don't need to worry about that one.

File: src/diffusers/training_utils.py (hunk 4 above)
Comment (patil-suraj) [reply src/diffusers/training_utils.py:95]: if `use_ema_warmup` is `True` then it'll compute the decay as per the previous `EMAModel`, if it's `False` (which is the default, it will compute it the same way we were using for SD. Does that makes sense ?

File: src/diffusers/training_utils.py (hunk 6 above)
Comment (patrickvonplaten) [reply src/diffusers/training_utils.py:75]: Think if we can avoid a breaking change, let's try to do it though no? Should be relatively easy by doing:
```suggestion
        if issubclass(parameters, torch.nn.Module):
            deprecate(...)
            parameters = parameters.parameters()
            
        parameters = list(parameters)
        self.shadow_params = [p.clone().detach() for p in parameters]
```

Comment (patrickvonplaten) [reply src/diffusers/training_utils.py:75]: That should catch 99% of the breaking use cases no? We could also add a kwargs.pop("model") to be super exact

File: src/diffusers/training_utils.py [hunk 10]
```diff
@@ -39,88 +40,173 @@ def set_seed(seed: int):
...
-            try:
-                ema_param = ema_params[key]
-            except KeyError:
-                ema_param = param.float().clone() if param.ndim == 1 else copy.deepcopy(param)
-                ema_params[key] = ema_param
-
-            if not param.requires_grad:
-                ema_params[key].copy_(param.to(dtype=ema_param.dtype).data)
-                ema_param = ema_params[key]
+    def step(self, parameters):
```
Comment (patrickvonplaten) [inline src/diffusers/training_utils.py:103]: Would also try to deprecate the `parameters` here is it should be quite simple

**Review by patrickvonplaten — COMMENTED** [review]
The PR looks very nice to me! I'd also just try to make it more or less 100% backwards compatible (let's try to be role-models when it comes to that in OSS).

Think it's not too hard:
- detect if a model is passed
- probs have to keep a mapping of parameters to names to be able to use new logic but also return model as done previously 
- we can relatively quickly deprecate here then I think

File: src/diffusers/training_utils.py (hunk 10 above)
Comment (patil-suraj) [reply src/diffusers/training_utils.py:103]: Do you mean deprecate `model`? `parameters` are required for `step`.

**Comment by patil-suraj** [comment]
The `EMAModel` should now be 100% backwards compatible. @patrickvonplaten @pcuenca would be nice if you could take one more look.

**Comment by patil-suraj** [comment]
The failing tests are unrelated.

File: src/diffusers/training_utils.py [hunk 11]
```diff
@@ -39,88 +42,232 @@ def set_seed(seed: int):
...
+
+        self.min_decay = state_dict.get("min_decay", self.min_decay)
+        if not isinstance(self.min_decay, float):
+            raise ValueError("Invalid min_decay")
+
+        self.optimization_step = state_dict.get("optimization_step", self.optimization_step)
+        if not isinstance(self.optimization_step, int):
+            raise ValueError("Invalid optimization_step")
+
+        self.update_after_step = state_dict.get("update_after_step", self.update_after_step)
```
Comment (pcuenca) [inline src/diffusers/training_utils.py:244]: 👍

**Review by pcuenca — APPROVED** [review]
Looks great! Just pointed out a couple nits.

File: src/diffusers/training_utils.py [hunk 12]
```diff
@@ -39,88 +42,232 @@ def set_seed(seed: int):
...
+            deprecate("min_value", "1.0.0", deprecation_message, standard_warn=False)
+            min_decay = kwargs["min_value"]
+
+        parameters = list(parameters)
+        self.shadow_params = [p.clone().detach() for p in parameters]
 
         if device is not None:
-            self.averaged_model = self.averaged_model.to(device=device)
+            self.shadow_params = [
+                p.to(device=device) if p.is_floating_point() else p.to(device=device) for p in self.shadow_params
```
[the remark this answers is not in the document]
Comment (patil-suraj) [reply src/diffusers/training_utils.py:113]: Good catch Pedro! We could just leverage the `.to` method here!

[the remark this answers is not in the document]
Comment (patil-suraj) [reply src/diffusers/training_utils.py]: Also, this is for backwards compatibility, like `torch` the user should call `.to` to put the `EMAModel` on the device, will actually depricate the `device` arg.
